File: Problems/Graph/Day-11/sol/Prajjawal Vaishya/Solution1.cpp
#include <iostream>
#include <vector>


/*
Problem: Kefa and Park (Codeforces 580C)

Short Statement:
Given a rooted tree with vertices marked as having a cat or not, find the number of leaf nodes (restaurants) 
reachable from the root such that the path contains no more than 'm' consecutive vertices with cats.

Detailed Approach:
1. We represent the tree using an adjacency list.
2. We perform a Depth First Search (DFS) starting from the root (vertex 1).
3. During the traversal, we maintain a variable `consecutiveCats` for the current path:
   - If the current node has a cat, we increment `consecutiveCats`.
   - If the current node does not have a cat, we reset `consecutiveCats` to 0 (since the sequence is broken).
4. Pruning Condition: If `consecutiveCats` exceeds 'm' at any point, we stop traversing that branch immediately 
   because any restaurant down that path is unreachable.
5. Leaf Check: If we reach a node that has no children (except its parent) and the path is valid, we increment our answer.

Time Complexity: O(N) - We visit each node and edge exactly once.
Space Complexity: O(N) - Used for the adjacency list and the recursion stack.

// Submission Link: https://codeforces.com/contest/580/submission/357205990
*/

using namespace std;
const int MAXN = 100005;
vector<int> adj[MAXN];
int hasCat[MAXN];
int n, m;
int totalRestaurants = 0;
void dfs(int u, int p, int consecutiveCats) {
    if (hasCat[u]) {
        consecutiveCats++;
    } else {
        consecutiveCats = 0;
    }
    if (consecutiveCats > m) {
        return;
    }
    bool isLeaf = true;
    for (int v : adj[u]) {
        if (v != p) {
            isLeaf = false;
            dfs(v, u, consecutiveCats);
        }
    }
    if (isLeaf) {
        totalRestaurants++;
    }
}


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    if (!(cin >> n >> m)) return 0;
    for (int i = 1; i <= n; i++) cin >> hasCat[i];

    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    dfs(1, -1, 0);
    cout << totalRestaurants << endl;
    return 0;
}

File: Problems/Graph/Day-11/sol/Chithra/Solution1.java
/*
Problem: C. Kefa and Park

Approach:
- Represent the park as a tree using an adjacency list.
- Use DFS starting from the root (node 1).
- Track the number of consecutive vertices with cats on the current path.
- If the count exceeds m, stop exploring that path.
- Count leaf nodes (restaurants) that satisfy the condition.

Time Complexity: O(n)
Space Complexity: O(n)

Submission link: https://codeforces.com/contest/769/submission/360849037
*/
import java.io.*;
import java.util.*;
 
public class Main {
    static int n, m;
    static int[] cat;
    static List<Integer>[] tree;
    static int answer = 0;
 
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
 
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
 
        cat = new int[n + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            cat[i] = Integer.parseInt(st.nextToken());
        }
 
        tree = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            tree[i] = new ArrayList<>();
        }
 
        for (int i = 0; i < n - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            tree[u].add(v);
            tree[v].add(u);
        }
 
        dfs(1, 0, 0);
 
        System.out.println(answer);
    }
 
    static void dfs(int node, int parent, int consecutiveCats) {
        if (cat[node] == 1) {
            consecutiveCats++;
        } else {
            consecutiveCats = 0;
        }
        if (consecutiveCats > m) return;
 
        boolean isLeaf = true;
 
        for (int next : tree[node]) {
            if (next != parent) {
                isLeaf = false;
                dfs(next, node, consecutiveCats);
            }
        }
 
        if (isLeaf) {
            answer++;
        }
    }
}